### PR TITLE
Ttt 23 create the orchestrator for tic tac toe

### PR DESCRIPTION
**Pull request by billlanyon**
Pull request as requested sir.

### CODE REVIEW

File: com/bl/ttt/main.py [hunk 1]
```diff
@@ -1,43 +1,73 @@
...
+get_next_move = True
+
+
+def process_another_move(game):
+    player_id, cell = input("Please enter player X or O, a space, and then a cell from 0 to 8: ").split(' ')
+    player_id = player_id.upper()
+    cell = int(cell)
+    if player_id is not None and (player_id == 'O' or player_id == 'X'):
+        if cell is not None and 0 <= cell <= 8:
+            move = TictactoeMove(player_id, cell)
```
Comment (BasilBibi) [inline com/bl/ttt/main.py]: Seems like some of the validation has crept into the orchestrator ?

File: com/bl/ttt/main.py [hunk 2]
```diff
@@ -1,43 +1,73 @@
...
+            if game.is_valid_move(move):
+                game.make_valid_move(move)
+
+                if game.has_won(player_id):
+                    print(game)
+                    print(f'Player {player_id} has won.')
+                    return game_over
+                elif game.is_draw():
```
Comment (BasilBibi) [inline com/bl/ttt/main.py]: Interesting that you added an is_draw - cool !

File: com/bl/ttt/tictactoe.py [hunk 3]
```diff
@@ -53,6 +56,10 @@ def _is_any_diagonal_complete(self, player_id):
                 self.cells[6] == self.cells[4] == self.cells[2] == player_id:
             return True
 
+    def is_draw(self):
+        if ' ' not in self.cells:
```
Comment (BasilBibi) [inline com/bl/ttt/tictactoe.py]: how about 
`return ' ' not in self.cells`

File: tests/com/bl/ttt/test_Tictactoe_is_draw.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+               'X', 'O', 'X',
+               'O', 'X', 'O']
+    assert g.is_draw()
+
+
+def test_02_is_draw_identifies_not_yet_draw():
+    g = Tictactoe()
+    g.cells = ['X', 'O', 'X',
+               'X', 'O', ' ',
+               'O', 'X', 'O']
```
Comment (BasilBibi) [inline tests/com/bl/ttt/test_Tictactoe_is_draw.py:16]: What happens when we have a winner and there are still ' ' cells ?

File: com/bl/ttt/tictactoe.py (hunk 3 above)
Comment (billlanyon) [reply com/bl/ttt/tictactoe.py]: Done

File: tests/com/bl/ttt/test_Tictactoe_is_draw.py (hunk 4 above)
Comment (billlanyon) [reply tests/com/bl/ttt/test_Tictactoe_is_draw.py:16]: test_03_is_draw_identifies_win_not_draw() added

File: com/bl/ttt/main.py (hunk 1 above)
Comment (billlanyon) [reply com/bl/ttt/main.py]: This seemed to be a minimum to reprompt for valid input against the is_valid_move method: should it be removed?

**Comment by billlanyon** [comment]
Game runs in command line until players no longer want to play:
CL1 2020-06-22 at 14 02 55
CL2 2020-06-22 at 14 03 47
CL3 2020-06-22 at 14 06 10

File: com/bl/ttt/main.py (hunk 1 above)
Comment (BasilBibi) [reply com/bl/ttt/main.py]: That puts validation logic in two places - does that seem right to you ?

Comment (billlanyon) [reply com/bl/ttt/main.py]: No, on reflection is does not.

**Review by BasilBibi — APPROVED** [review]

**Pull request merged**